mcstatus/status_response.py

```diff
@@ -0,0 +1,640 @@
+        "aqua": "b",
+        "dark_aqua": "3",
+        "dark_blue": "1",
+        "blue": "9",
+        "light_purple": "d",
+        "dark_purple": "5",
+        "white": "f",
+        "gray": "7",
+        "dark_gray": "8",
+        "black": "0",
+    },
+}
+
+
+def _validate_data(raw: Dict[str, Any], who: str, required: Iterable[Tuple[str, type]]) -> None:
+    """Ensure that all required keys are present, and have the specified type.
+
+    :param raw: The raw dict answer to check.
+    :param who: The name of the object that is checking the data. Example `status`, `player` etc.
+    :param required: An iterable of string and type. The string is the required key which must be in `raw`, and
+        the `type` is the type that the key must be. If you want to ignore check of the type, set the type to `object`.
+    :raises ValueError: If the required keys are not present.
+    :raises TypeError: If the required keys are not of the specified type.
+    """
+    for required_key, required_type in required:
+        if required_key not in raw:
+            raise ValueError(f"Invalid {who} object (no '{required_key}' value)")
+        if not isinstance(raw[required_key], required_type):
+            raise TypeError(
+                f"Invalid {who} object (expected '{required_key}' to be {required_type}, was {type(raw[required_key])})"
+            )
+
+
+@dataclass
+class AbstractDataclass(ABC):
+    """Abstract base class for all dataclasses.
+
+    We can't use just `ABC` because we need raise TypeError if we try to
+    instantiate an abstract dataclass. Class which inherit from `ABC` raises
+    it only if it has abstract methods.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        """Check if someone tries to instantiate an abstract class.
+
+        :raises NotImplementedError: If someone tries to instantiate an abstract class.
+        :return: The new instance of child abstract class.
+        """
+        if cls is AbstractDataclass or cls.__bases__[0] is AbstractDataclass:
+            raise NotImplementedError("Cannot instantiate abstract class.")
+        return super().__new__(cls)
+
+
+@dataclass
+class MCStatusResponse(AbstractDataclass):
+    """Class for storing shared data from a status response.
+
+    :param motd: Message Of The Day. Also known as `Description`.
+    :param latency: Latency between a server and the client (you). In milliseconds.
+    """
+
+    players: StatusPlayers
+    version: StatusVersion
+    motd: str
+    latency: float
+
+    @classmethod
+    @abstractmethod
+    def build(cls, *args, **kwargs) -> Self:
+        """Build MCStatusResponse and check is it valid.
+
+        :param args: Arguments in specific realisation.
+        :param kwargs: Keyword arguments in specific realisation.
+        :return: `MCStatusResponse` object.
+        """
+        raise NotImplementedError("You can't use abstract methods.")
+
+
+@dataclass
+class NewJavaStatusResponse(MCStatusResponse):
```

Comment:
Don't call these classes `New...`, they should simply be `JavaStatusResponse`, and the deprecated classes should then be `DeprecatedJavaStatusResponse`, which can then be imported from `pinger.py`/`bedrock_status.py` with 
```py
from mcstatus.status_response import DeprecatedJavaStatusResponse as PingResponse
```
which can then be easily removed after the deprecation period and simply replaced by
```py
from mcstatus.status_response import JavaStatusResponse
```

This is needed because some people may want to use the new classes introduced here in their code directly by importing them. Keeping the classes prefixed with `New` would mean that after the deprecation period, we would have to keep using the `New` prefixed classes, even though there won't be any old/original classes, and so that name wouldn't make any sense.

Reply:
Only applicable if the entire class won't be moved into their respective files as pointed out in another review. Moving these deprecated classes to their original places makes much more sense imo, but if there is something preventing that, this should be done instead.

Reply:
My idea about deprecated classes is:
- On 2022-08 we need only remove code in `status_response.py`.
- This code can be easily remove, we don't need to find usages and remove them too.
- Deprecated classes can be used as old, same as new. No need to switch between them, so no-one will change their code after 2022-08 (if they merged once).
- We don't need create second breaking changes update.

---

And let's see what my code now implementing:
- [ ] On 2022-08 we need only remove code in `status_response.py`.

We need only remove all under N string, plus remove unused imports in `status_response.py`, and rename `New` classes.
P.S. Oops, also we need remove aliases in old places.
- [x] This code can be easily remove, we don't need to find usages and remove them too.

Only few steps (see previous answer).
- [x] Deprecated classes can be used as old, same as new. No need to switch between them, so no-one will change their code after 2022-08 (if they updated once).

Overloads was implemented on conflicting methods, old interface supporting same as new. No need to switch between them.
- [x] We don't need create second breaking changes update.

Classes prefixed with `New` **NOT A PART** of a public API (because not listed in `__all__`), so no-one should use it. There is no breaking changes, if you delete variable with `__` prefix.

---

What suggest you:
- [ ] On 2022-08 we need only remove code in `status_response.py`.

You need remove some imports and code usage in other files.
- [ ] This code can be easily remove, we don't need to find usages and remove them too.

See previous.
- [ ] Deprecated classes can be used as old, same as new. No need to switch between them, so no-one will change their code after 2022-08 (if they merged once).

User need to switch between classes, because they are a part of public API.
- [ ] We don't need create second breaking changes update.

You need deprecate switch method. Example argument in `status` method, or anything else.

Reply:
Actually, you're probably right, there may still be some better way to handle this but I can't think of any atm.